gt/main.c

```diff
@@ -869,14 +793,381 @@ print_unsent_policy(struct ggu_policy *policy)
+	struct gt_instance *instance, unsigned int *index)
+{
+	unsigned int i;
+	for (i = 0; i < instance->num_ggu_pkts; i++) {
+		if (pkt_info->outer_ethertype != instance->ggu_pkts[i].proto)
+			continue;
+
+		if (instance->ggu_pkts[i].proto == ETHER_TYPE_IPv4) {
+			struct ipv4_hdr *ipv4_hdr =
+				(struct ipv4_hdr *)pkt_info->outer_l3_hdr;
+			if (instance->ggu_pkts[i].ip.v4.s_addr ==
+					ipv4_hdr->src_addr) {
+				*index = i;
+				return &instance->ggu_pkts[i];
+			}
+		} else if (likely(instance->ggu_pkts[i].proto ==
+				ETHER_TYPE_IPv6)) {
+			struct ipv6_hdr *ipv6_hdr =
+				(struct ipv6_hdr *)pkt_info->outer_l3_hdr;
+			if (ipv6_addrs_equal(ipv6_hdr->src_addr,
+					instance->ggu_pkts[i].ip.v6.s6_addr)) {
+				*index = i;
+				return &instance->ggu_pkts[i];
+			}
+		}
+	}
+	return NULL;
+}
+
+/* Send the notification packet at @index. */
+static void
+send_notify_pkt(struct gt_instance *instance, unsigned int index)
+{
+	struct ggu_notify_pkt *ggu_pkt;
+	struct udp_hdr *notify_udp;
+	unsigned int lcore_id = rte_lcore_id();
+
+	if (instance->num_ggu_pkts < index) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gt: instance on lcore %u was asked to send a notification packet at index %u but only %u exist\n",
+			lcore_id, index, instance->num_ggu_pkts);
+		return;
+	}
+
+	ggu_pkt = &instance->ggu_pkts[index];
+
+	/*
+	 * Complete packet fields that can only be done
+	 * when packet is ready to be transmitted.
+	 */
+
+	if (ggu_pkt->proto == ETHER_TYPE_IPv4) {
+		struct ipv4_hdr *notify_ipv4 =
+			rte_pktmbuf_mtod_offset(ggu_pkt->buf,
+				struct ipv4_hdr *, ggu_pkt->buf->l2_len);
+		notify_ipv4->total_length = rte_cpu_to_be_16(
+			ggu_pkt->buf->data_len - ggu_pkt->buf->l2_len);
+
+		/* Offload the UDP checksum. */
+		notify_udp = (struct udp_hdr *)&notify_ipv4[1];
+		notify_udp->dgram_cksum = rte_ipv4_phdr_cksum(notify_ipv4,
+			ggu_pkt->buf->ol_flags);
+	} else if (likely(ggu_pkt->proto == ETHER_TYPE_IPv6)) {
+		struct ipv6_hdr *notify_ipv6 =
+			rte_pktmbuf_mtod_offset(ggu_pkt->buf,
+				struct ipv6_hdr *, ggu_pkt->buf->l2_len);
+		notify_ipv6->payload_len = rte_cpu_to_be_16(
+			ggu_pkt->buf->data_len - ggu_pkt->buf->l2_len -
+			sizeof(struct ipv6_hdr));
+
+		/* Offload the UDP checksum. */
+		notify_udp = (struct udp_hdr *)&notify_ipv6[1];
+		notify_udp->dgram_cksum = rte_ipv6_phdr_cksum(notify_ipv6,
+			ggu_pkt->buf->ol_flags);
+	} else {
+		rte_panic("Unexpected condition: gt at lcore %u sending notification packet to Gatekeeper server with unknown IP version %hu\n",
+			lcore_id, ggu_pkt->proto);
+	}
+
+	notify_udp->dgram_len =
+		rte_cpu_to_be_16((uint16_t)(ggu_pkt->buf->data_len -
+			ggu_pkt->buf->l2_len - ggu_pkt->buf->l3_len));
+
+	if (rte_eth_tx_burst(get_net_conf()->front.id, instance->tx_queue,
+			&ggu_pkt->buf, 1) != 1) {
+		print_unsent_policies(ggu_pkt);
+		rte_pktmbuf_free(ggu_pkt->buf);
+	}
+
+	/* Move all remaining buffered packets over to the left. */
+	rte_memcpy(&instance->ggu_pkts[index], &instance->ggu_pkts[index + 1],
```

Comment:
This is on the critical path, the memory copy here could consume a lot of CPU cycles. Maybe we can find ways to optimize it.